app/services/attachments/concatenation_service.rb

```diff
@@ -0,0 +1,153 @@
+              reverse_reads << attachment
+            end
+          end
+          concatenate_paired_end_reads(forward_reads, reverse_reads)
+        elsif is_single_end
+          concatenate_single_end_reads(attachments)
+        end
+
+        # if option is selected then destroy the original files
+        attachments.each(&:destroy!) if concatenation_params[:delete_originals]
+      end
+    rescue Attachments::ConcatenationService::AttachmentConcatenationError => e
+      attachable.errors.add(:base, e.message)
+      attachable
+    end
+
+    private
+
+    # Validates attachments to ensure they are of the correct type.
+    # Files can only be concatenated as follows. No mixing of types:
+    # Paired-end -> Paired-end
+    # Single-end -> Single-end
+    def validate_same_types(attachments, is_paired_end, is_single_end) # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
+      valid_type = true
+
+      attachments.each do |attachment|
+        if is_single_end
+          valid_type = !attachment.metadata.key?('type')
+        elsif is_paired_end
+          valid_type = attachment.metadata.key?('type')
+        end
+      end
+
+      if valid_type
+        expected_extension = attachments.first.file.filename.to_s.partition('.').last
+        attachments.each do |attachment|
+          extension = attachment.file.filename.to_s.partition('.').last
+          next unless extension != expected_extension
+
+          valid_type = false
+          raise AttachmentConcatenationError,
+                I18n.t('services.attachments.concatenation.incorrect_fastq_file_types')
+        end
+      end
+
+      return if valid_type
+
+      raise AttachmentConcatenationError,
+            I18n.t('services.attachments.concatenation.incorrect_file_types')
+    end
+
+    # Concatenates the single end reads into a single-end file
+    def concatenate_single_end_reads(attachments)
+      basename = concatenation_params[:basename] || 'concatenated_file'
+      extension = attachments.first.file.filename.to_s.partition('.').last
+
+      extension = 'fastq.gz' if extension == 'gz'
+
+      blobs = []
+      attachments.each do |attachment|
+        blobs << attachment.file.blob
+      end
+
+      composed_blob = ActiveStorage::Blob.compose(blobs, filename: "#{basename}_S1_L001_R1_001.#{extension}")
+
+      Attachments::CreateService.new(current_user, attachable, { files: [composed_blob.signed_id] }).execute
+    end
+
+    # Concatenates the paired-end reads into a multiple paired-end files
+    def concatenate_paired_end_reads(forward_reads, reverse_reads) # rubocop:disable Metrics/MethodLength, Metrics/AbcSize
+      basename = concatenation_params[:basename] || 'concatenated_file'
+      extension = forward_reads.first.file.filename.to_s.partition('.').last
+
+      extension = 'fastq.gz' if extension == 'gz'
+
+      files = []
+      if forward_reads.length.positive?
```

Comment:
**nitpick (blocking):** Due to other comments I think we should remove the `positive?` checks and I think it would also be good to refactor the contents of the `if` and other `if` to be a new method, since this logic is shared between `concatenate_single_end_reads` and this method.

Reply:
Refactored the logic to remove the `positive?` checks and moved common code between this method and the one for single end reads into a new method in 349d6987f0f58fb6c2481ed054fe66446e6a6090